### PR TITLE
TearSheet header height change fix

### PR DESCRIPTION
**Pull request by hectordavis**
**Summary**
- This fixes an issue where if a content were dynamically added to the TearSheet header, the TearSheet content would expand beyond its container and get cropped by the bottom of the viewport.

**Acceptance Test (how to verify the PR)**
- Check the component story. I've added buttons on the tear sheets to include additional content to their header. The tear sheet size should be readjusted to prevent it from being cropped.

**Regression Test (how to make sure this PR doesn't break old functionality)**
The tests below can be checked via story
- Check the stack feature (one tear sheet on top of the other) to see if they still behave like before. 
- Check if the tear sheet's whole content is displayed

### CODE REVIEW

File: packages/react/src/components/TearSheet/TearSheetWrapper.jsx [hunk 1]
```diff
@@ -48,6 +49,32 @@ const TearSheetWrapper = ({ isOpen, className, onCloseAllTearSheets, children })
...
+  const tearSheetsNodes = containers.reduce(
+    (acc, curr) => [...acc, curr?.current?.children[0]],
+    []
+  );
+
+  useEffect(() => {
+    if (
+      activeTearSheetIdx !== null &&
+      !Number.isNaN(
+        Math.round(
```
Comment (blechner) [inline packages/react/src/components/TearSheet/TearSheetWrapper.jsx:61]: Maybe `Math.round(tearSheetsNodes?.[activeTearSheetIdx]?.children?.[0].getBoundingClientRect().height)` could be a variable.

Comment (hectordavis) [reply packages/react/src/components/TearSheet/TearSheetWrapper.jsx]: Good catch. Thanks!

**Review by blechner — APPROVED** [review]

File: packages/react/src/components/TearSheet/TearSheetWrapper.jsx [hunk 2]
```diff
@@ -48,6 +49,24 @@ const TearSheetWrapper = ({ isOpen, className, onCloseAllTearSheets, children })
...
   );
 
+  const tearSheetsNodes = containers.reduce(
+    (acc, curr) => [...acc, curr?.current?.children[0]],
+    []
+  );
+
+  useEffect(() => {
+    const tearSheetHeaderHeight = Math.round(
+      tearSheetsNodes?.[activeTearSheetIdx]?.children?.[0].getBoundingClientRect().height
```
Comment (kevinsperrine) [inline packages/react/src/components/TearSheet/TearSheetWrapper.jsx:59]: This probably makes more sense as a useLayoutEffect since you're calling getBoundingClientRect for measurements. https://kentcdodds.com/blog/useeffect-vs-uselayouteffect

**Review by kevinsperrine — APPROVED** [review]
just the one note on useLayoutEffect.

**Comment by davidicus** [comment]
I am not understanding the reason to calculate the height of the header. Can't we have two items in a flex container that can grow with their content? Like this example? https://codesandbox.io/s/green-http-zq8jr

**Review by davidicus — CHANGES_REQUESTED** [review]
Waiting on reply

**Comment by hectordavis** [comment]
Hey, @davidicus 
The tear sheet has a fixed position. If something changes the header height, that's going to affect the rest of the tear sheet because the `--content-max-height` value won't be changed. Now, if we change the window size for instance, this would be recalculated and the tear sheet would adjust, but a change in the header's height won't trigger that. At least it didn't before this PR 🙂 

The tear sheet has a top limit and a bottom limit. The header is taken into consideration as well to decide the max-height of the **content**. After the calculations to get this max-height are done, we need something to trigger them again if something changes (other than what we already checked in the useEffect, like the active tear sheet index, window size, the initial container width, etc.)

The header can't grow upwards, so it'll force the content down causing the tear sheet to be cropped. That's why we need to recalculate the content max-height when the header changes.

It's also worth noting that there's also the case of stacked tear sheets, so they need to keep the same difference in position from each other (top and bottom) even if their headers change dynamically.

In short, I'd say the main point is the fixed positioning.

**Comment by davidicus** [comment]
> Hey, @davidicus
> The tear sheet has a fixed position. If something changes the header's height, that's going to push down the rest of the tear sheet because the `--content-max-height` value won't be recalculated. Now, if we change the window size for instance, then the tear sheet would adjust due to the existing useEffect, but a change in the header's height won't trigger that. At least it didn't before this PR 🙂
> 
> The tear sheet has a top limit and a bottom limit. The header is taken into consideration as well to decide the max-height of the **content** below. After the calculations to get this max-height are done, we need something to trigger them again if something changes (other than what we already checked in the useEffect, like the active tear sheet index, window size, the initial container width, etc.)
> 
> The header can't grow upwards, so it'll force the content down causing the tear sheet to be cropped. That's why we need to recalculate the content max-height if the header is different in size than it was before. Let's pick a TearSheet with a height of 700px (header: 100px, content: 600px). If we add a 50px high content in the header, the expected would be: `header: 150px, content: 550px`, but without the calculation, we have: `header: 150px, content: 600px` resulting in a TearSheet 750px high (with 50px cropped at the bottom).
> 
> It's also worth noting that there's also the case of stacked tear sheets, so they need to keep the same difference in position from each other (top and bottom) even if their headers change dynamically.
> 
> One thing that might come to mind about all of this is: "Why not just use `height: 100%` and / or `max-height: 100%`... Well, it doesn't work. We need the "fixed" `--content-max-height` value to set the tear sheet content to the maximum allowed to work as the design documents and to keep the style when stacking, etc...
> 
> In short, I'd say the main point is the fixed positioning + the tear sheet's content max-height.

@hectordavis I think this can be achieved with styles. The spec is saying tear sheets always are 88px from top of screen and go to bottom of screen. I think the layout and placement can be achieved with styles alone. I may be missing something so perhaps we can hop on a  call.

**Comment by hectordavis** [comment]
Closing this PR as the issue was fixed by PR #2400
